modules/auth/ldap/ldap.go

```diff
@@ -201,6 +218,96 @@ func checkRestricted(l *ldap.Conn, ls *Source, userDN string) bool {
 	return false
 }
 
+// CheckGroupFilter :
+func (ls *Source) CheckGroupFilter(l *ldap.Conn, groupSR *ldap.SearchResult, filter string) bool {
+	for _, groupEntry := range groupSR.Entries {
+		search := ldap.NewSearchRequest(groupEntry.DN, ldap.ScopeBaseObject, ldap.NeverDerefAliases, 0, 0, false, filter, []string{}, nil)
+		sr, err := l.Search(search)
+		if (err == nil) && (len(sr.Entries) > 0) {
+			return true
+		}
+	}
+	log.Trace("LDAP group search with filter %v found no matching entries", filter)
+	return false
+}
+
+// ProcessUserEntry :
+func (ls *Source) ProcessUserEntry(entry *ldap.Entry, l *ldap.Conn) *SearchResult {
+	username := entry.GetAttributeValue(ls.AttributeUsername)
+	firstname := entry.GetAttributeValue(ls.AttributeName)
+	surname := entry.GetAttributeValue(ls.AttributeSurname)
+	mail := entry.GetAttributeValue(ls.AttributeMail)
+
+	var sshPublicKey []string
+	if len(strings.TrimSpace(ls.AttributeSSHPublicKey)) > 0 {
+		sshPublicKey = entry.GetAttributeValues(ls.AttributeSSHPublicKey)
+	}
+
+	var hasAdminGroup = false
+	var hasRestrictedGroup = false
+	if len(strings.TrimSpace(ls.GroupSearchBase)) > 0 && len(strings.TrimSpace(ls.GroupSearchFilter)) > 0 {
+		var groupUID string
+		if len(strings.TrimSpace(ls.UserAttributeInGroup)) > 0 {
+			groupUID = entry.GetAttributeValue(ls.UserAttributeInGroup)
+		} else {
+			groupUID = entry.DN
+		}
+		log.Trace("User attribute used in LDAP group: %v", groupUID)
+
+		groupFilter, ok := ls.sanitizedGroupQuery(groupUID)
+		if !ok {
+			return nil
+		}
+
+		groupSearch := ldap.NewSearchRequest(
+			ls.GroupSearchBase, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false, groupFilter, []string{}, nil)
+
+		sr, err := l.Search(groupSearch)
+		if err != nil {
+			log.Error("LDAP group search failed unexpectedly! (%v)", err)
+			return nil
+		}
+
+		if len(strings.TrimSpace(ls.MemberGroupFilter)) > 0 {
+			if !ls.CheckGroupFilter(l, sr, ls.MemberGroupFilter) {
+				log.Error("No group matched the required member group filter!")
+				return nil
+			}
+		}
+
+		if len(strings.TrimSpace(ls.AdminGroupFilter)) > 0 {
+			hasAdminGroup = ls.CheckGroupFilter(l, sr, ls.AdminGroupFilter)
+			if hasAdminGroup {
+				log.Info("LDAP user %s is in admin group!", username)
+			}
+		}
+
+		if len(strings.TrimSpace(ls.RestrictedGroupFilter)) > 0 {
+			hasRestrictedGroup = ls.CheckGroupFilter(l, sr, ls.RestrictedFilter)
```

Comment:
shouldn't it be `ls.RestrictedGroupFilter` instead of `ls.RestrictedFilter` ?
```suggestion
			hasRestrictedGroup = ls.CheckGroupFilter(l, sr, ls.RestrictedGroupFilter)
```